doc/website/advanced/best-practice-for-testing.md

```diff
@@ -0,0 +1,257 @@
+
+## How To Do It Better
+
+At first, the test is split into multiple test cases according the AAA pattern
+
+```c++
+TEST_F(SingleDigitNumber_test, DefaultConstructedObjectIsCorrectlyInitialized)
+{
+    constexpr uint64_t EXPECTED_VALUE{0U};
+
+    SingleDigitNumber sut;
+
+    EXPECT_EQ(static_cast<uint64_t>(sut), EXPECTED_VALUE);
+}
+```
+
+The test also has a meaningful name. If this fails in the CI, it is immediately clear what is broken.
+Additionally, the tested object is called `sut`. This makes it easy to identify the actual test object.
+Lastly, a `constexpr` is used for the expected value. This removes a magic value and also makes the output of a failed test more readable, since it is immediately clear what's the actual tested value and what's the expected value.
+
+Now let's continue with further tests, applying the ZOMBIES principle
+
+```c++
+TEST_F(SingleDigitNumber_test, ConstructionWithValidValueCreatesNumberWithSameValue)
+{
+    constexpr uint64_t NUMBER_VALUE{7U};
+    constexpr uint64_t EXPECTED_VALUE{NUMBER_VALUE};
+
+    SingleDigitNumber sut{NUMBER_VALUE};
+
+    EXPECT_EQ(static_cast<uint64_t>(sut), EXPECTED_VALUE);
+}
+
+TEST_F(SingleDigitNumber_test, ConstructionWithInvalidValueResultsInSaturation)
+{
+    constexpr uint64_t NUMBER_VALUE{42U};
+    constexpr uint64_t EXPECTED_VALUE{9U};
+
+    SingleDigitNumber sut{NUMBER_VALUE};
+
+    EXPECT_EQ(static_cast<uint64_t>(sut), EXPECTED_VALUE);
+}
+
+TEST_F(SingleDigitNumber_test, AddingZeroDoesNotChangeTheNumber)
+{
+    constexpr SingleDigitNumber NUMBER{3U};
+    constexpr SingleDigitNumber ZERO{0U};
+    constexpr uint64_t EXPECTED_VALUE{3U};
+
+    auto sut = NUMBER + ZERO;
+
+    EXPECT_EQ(static_cast<uint64_t>(sut), EXPECTED_VALUE);
+}
+
+TEST_F(SingleDigitNumber_test, AddingOneIncreasesTheNumberByOne)
+{
+    constexpr SingleDigitNumber NUMBER{3U};
+    constexpr SingleDigitNumber ONE{1U};
+    constexpr uint64_t EXPECTED_VALUE{4U};
+
+    auto sut = NUMBER + ONE;
+
+    EXPECT_EQ(static_cast<uint64_t>(sut), EXPECTED_VALUE);
+}
+
+// and so on
+```
+
+These are some examples showing how to apply the ZOMBIES principle to find good test cases.
+They also exert all the good practices mentioned previously, like clear and distinct names or avoiding magic numbers.
+
+# Slightly More Advanced Topics
+
+## Typed Tests
+
+TODO:
+- simple example which can be used as template
+
+## Parameterized Tests
+
+TODO:
+- simple example which can be used as template
+- https://www.sandordargo.com/blog/2019/04/24/parameterized-testing-with-gtest
+- use `std::tie`
+
+## Mocks
+
+TODO:
+- just a few links
+
+# Conclusion
+
+- apply the AAA pattern to structure the test and check only one aspect of the code at a time
+- use the ZOMBIES principle to find sensible test cases
+- meaningful names for the tests to indicate what the test is supposed to do
```

Comment:
I'd be more explicit here. The test name should indicate what's done in the test (or going in) and include the outcome of the test, e.g. "...IsSuccessful" or "...ReturnsXYZ".

Reply:
I'm a bit hesitant because I don't want to put us in a straitjacket. How about "meaningful names for the tests to indicate what the test is supposed to do and the outcome" and adding this as examples but not limiting it to this?

Reply:
I am with @elBoberido here. Sometimes it is not easy to express the test intention with those Suffixes and the author needs some wiggle room.

Reply:
I am with @marthtz  here. Let's say you are the poor developer who has to fix a test named like `CallingMethodWithParameter`. Then you now what is tested here but you do not know what kind of outcome the author expected. Should this always work, should it fail?

Then you take a look inside the code see that it fails and the implementation fails but was it really the intention here. Maybe two things came together - an implementation error in the test and in the actual implementation?!

In the end I feel that test names are like tiny requirements which states DoingBlaLeadsToBehaviorFuu.